grafana/risingwave-dev-dashboard.dashboard.py

```diff
@@ -678,6 +678,46 @@ def section_streaming(outer_panels):
         outer_panels.row_collapsed(
             "Streaming",
             [
+                panels.timeseries_count(
+                    "Barrier Number",
+                    "The number of barriers that have been ingested but not completely processed. This metric reflects the "
+                    "current level of congestion within the system.",
+                    [
+                        panels.target(f"{metric('all_barrier_nums')}", "all_barrier"),
+                        panels.target(
+                            f"{metric('in_flight_barrier_nums')}", "in_flight_barrier"
+                        ),
+                    ],
+                ),
+                panels.timeseries_latency(
+                    "Barrier Latency",
+                    "The time that the data between two consecutive barriers gets fully processed, i.e. the computation "
+                    "results are made durable into materialized views or sink to external systems. This metric shows to users "
+                    "the freshness of materialized views.",
+                    quantile(
+                        lambda quantile, legend: panels.target(
+                            f"histogram_quantile({quantile}, sum(rate({metric('meta_barrier_duration_seconds_bucket')}[$__rate_interval])) by (le))",
+                            f"barrier_latency_p{legend}",
+                        ),
+                        [50, 90, 99, 999, "max"],
+                    )
+                    + [
+                        panels.target(
+                            f"rate({metric('meta_barrier_duration_seconds_sum')}[$__rate_interval]) / rate({metric('meta_barrier_duration_seconds_count')}[$__rate_interval])",
+                            "barrier_latency_avg",
+                        ),
+                    ],
+                ),
+                panels.timeseries(
+                    "Barrier pending time (secs)",
+                    "The duration since the last barrier has completed. "
+                    "Note that the metrics might be larger than the real value up to the Prometheus scrape interval.",
+                    [
+                        panels.target(
+                            f"time() - {metric('last_committed_barrier_time')}", "barrier_pending_time"
```

Comment:
It seems I completely misunderstood `time()`, and I should use `timestamp(last_committed_barrier_time)` instead

The changed version looks better

![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/JgO8A8bpio3B4YlQByok/8ece7930-b889-40e3-9d7d-151ecf39ef6b.png)

Actually I'm now very confused what does `time()` mean exactly (what is "evaluation time"?) And don't know what was I thinking. 🤡

> time() does not actually return the current time, but the time at which the expression is to be evaluated.

Reply:
> what is "evaluation time"?

It means the time when the `time()` function is evaluated by the query engine[1]. The Promlens demo gives a good demonstration: https://demo.promlens.com/?l=ITnOPpE8D0J.

I actually think it's a better fit here than `timestamp(last_committed_barrier_time)`, which is the timestamp of the sample, meaning it will fluctuate due to the scraping: https://demo.promlens.com/?l=fycmtZbOWVY

[1] Related code is at https://github.com/prometheus/prometheus/blob/c6a42f88915b4cd3076e93a3f8c5595a4ee934f1/promql/functions.go#L60-L64 
and https://github.com/prometheus/prometheus/blob/c6a42f88915b4cd3076e93a3f8c5595a4ee934f1/promql/engine.go#L1524

> How about recording the latency i.e. now()-last_committed_barrier_time when collecting?

Are we able to do that at scraping time? I'm a bit skeptical about it.

Reply:
> Isn't "the timestamp of the sample" = "scrape time"?

I don't quite follow. What I was saying is I'm not sure whether we can get calculate `now()-last_committed_barrier_time` when collecting (aka at scraping time), per @fuyufjh 's comment.


> What I don't understand is when is exprs evaluated? Is it scrape interval?

It's the query expression, it's evaluated when you send the query to the engine i.e. in most cases, when you open Grafana and Grafana sends the queries to the Prometheus backend.

Reply:
Sorry I typed too soon without checking: so when you issue a query to Prometheus, you need to specify the evaluation time  -- that means the time point you want to query from a a time series. Ans this evaluation time you specified will be the return value of `time()`. 

https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries

Reply:
I think what @xxchan would like to express in the panel is the duration between the time when a data point of `last_committed_barrier_time` is sampled and the data point value standing for when the last commit happened as latency. 

It should be a per data point value, so I think `time` won't fit here.

Reply:
However, I'm also a bit concerned of the accuracy of timestamp when the host is under high pressure, since the timestamp is assigned by the telemetry agent (a.k.a, grafana agent). I'm sure there will be a delta between the time a value of `last_committed_barrier_time` is generated by kernel and the time it's processed, especially when the scraped data is huge.

Reply:
> I think what @xxchan would like to express in the panel is the duration between the time when a data point of last_committed_barrier_time is sampled and the data point value standing for when the last commit happened as latency.

Why is this better than the evaluation time? `The time when a data point of last_committed_barrier_time is sampled` is exactly the cause of this Note:
```
Note that the metrics might be larger than the real value up to the Prometheus scrape interval.
```
And `time()` which basically is the timestamp value of the horizontal axis in the graph could eliminate that.

Reply:
Well, the yellow one is smoother than green one. Did I miss sth? @wjf3121

Reply:
Hmm.. that's a bit unexpected. Can you try plotting the following chart:
1. `time()`
2. `timestamp(last_committed_barrier_time)`
3. `last_committed_barrier_time`

I would expect 1 and 2 to look similar but 1 to be more smooth (strictly linear).
And 3 to be a bunch of flat segments of which values are increasing.

Reply:
> And time() which basically is the timestamp value of the horizontal axis

Thanks. This finally makes me understand `time()`. Previously I was thinking there's a value for each evaluation...  🤡

> 2. timestamp(last_committed_barrier_time)
> I would expect 1 and 2 to look similar

I think it's value is flat segments, with fixed spacing (scrap interval).

---

I think I finally understand the question:

The problem is: After datapoint `b1`, scrapped at `s1`, we don't know if we have a data point `b2` until `s2`.

So:
- `m1`: `timestamp(s1) - s1` (yellow line) is a **constant** between s1 and s2.
- `m2`: `time() - s1` (green line) keeps **increasing** between s1 and s2.

`time()` should be strictly larger than `timestamp(s1)` (the difference is just how long there's no data scrapped, so up to `scrape_interval`


<img width="755" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/37948597/3222231d-0ccb-40ff-9fc2-9f1454368882">

Reply:
> Can you try plotting the following chart:

Sure. I think it does make the distinction more clear

<img width="1255" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/37948597/3c3afb5e-0cc1-4ad5-bd71-72aa05a4935a">

Reply:
Yep. I think this may explain more: 

- The red line should be the "ideal" line. (barrier interval is 20).
- The green line is always above it (and sometimes overlap). Largest error is scrape interval (as I observed at the very beginning), when the barrier arrives right after a scrape. 
- The yellow line is always below it. The largest error is also scrape interval, when the barrier arrives right before a scrape. 

<img width="1001" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/37948597/3198c29b-f438-491a-8c47-77e8d7911995">

Reply:
So my conclusion is that the yellow one might be better as the alert.

Reply:
Yes exactly, by decreasing the scraping interval, `time() - last_committed_barrier_time` will get closer to the "ideal" one, but `timestamp(last_committed_barrier_time) - last_committed_barrier_time` is always something else -- it also reflects the freshness, just not as explanatory to me as using `time()`. That's why I would suggest using `time()` here.

Reply:
> So my conclusion is that the yellow one might be better as the alert.

I don't quite agree with this because alerting are usually tied with metrics collection, so scraping time will be taken into consideration when setting up a threshold.

I also don't think it matters too much in this case so let it be, YOLO.